src/api/paraswap/index.ts

```diff
@@ -0,0 +1,140 @@
+    params: ParaswapPriceQuoteParams & { chainId: NetworkID },
+    options?: Omit<QuoteOptions<true>, 'chainId'>
+  ): Promise<OptimalRate | null>
+  public async getQuote(
+    params: ParaswapPriceQuoteParams & { chainId: ParaswapCowswapNetworkID },
+    options?: Omit<QuoteOptions<false>, 'chainId'>
+  ): Promise<OptimalRate | null>
+  public async getQuote(
+    params: ParaswapPriceQuoteParams,
+    options: Omit<QuoteOptions<boolean>, 'chainId'> = { allowParaswapNetworks: false }
+  ): Promise<OptimalRate | null> {
+    const { chainId } = params
+    const { baseToken, quoteToken, fromDecimals, toDecimals, amount, kind, userAddress } = getValidParams(
+      params.chainId,
+      params
+    )
+
+    const paraSwap = this.getLib(chainId, options?.apiUrl || this.apiUrl)
+    if (!paraSwap) throw new CowError("ParaswapApi isn't compatible with chainId " + chainId)
+
+    // Buy/sell token and side (sell/buy)
+    const { sellToken, buyToken } = getTokensFromMarket({ baseToken, quoteToken, kind })
+    const swapSide = kind === OrderKind.BUY ? SwapSide.BUY : SwapSide.SELL
+
+    // https://developers.paraswap.network/api/get-rate-for-a-token-pair
+    const paraswapOptions: RateOptions = { ...this.rateOptions, ...options?.rateOptions }
+    // block any non-cow compatible chains
+    if (!options?.allowParaswapNetworks) {
+      const isCompatibleChain = getParaswapChainId(chainId)
+      if (!isCompatibleChain) {
+        error(
+          LOG_PREFIX,
+          `Chain id ${chainId} is not currently supported by the CoW contracts - please use: ${COMPATIBLE_PARASWAP_CHAINS_WITH_COW.join(
+            ','
+          )}. If you want to query all other ParaSwap specific chain ids, please pass { allowParaswapNetworks: true } as the query's "options" parameter.`
+        )
+
+        // return a null price
+        return null
+      }
+    }
+
+    debug(LOG_PREFIX, 'Getting price quote', params)
+
+    // Get price
+    const rateResult = await paraSwap.getRate(
+      sellToken,
+      buyToken,
+      amount,
+      userAddress,
+      swapSide,
+      paraswapOptions,
+      fromDecimals,
+      toDecimals
+    )
+
+    return handleResponse(rateResult)
+  }
+
+  public async getQuoteAllNetworks(
+    params: ParaswapPriceQuoteParams & { chainId: NetworkID },
+    options?: Omit<QuoteOptions<true>, 'chainId' | 'allowParaswapNetworks'>
+  ) {
+    return this.getQuote(params, { ...options, allowParaswapNetworks: true })
+  }
+
+  public async updateOptions(options: { apiUrl?: string; rateOptions?: RateOptions | null }) {
+    // null resets rateOptions to empty
+    if (options.rateOptions === null) {
+      this.rateOptions = {}
+    } else {
+      this.rateOptions = {
+        ...this.rateOptions,
+        ...(options.rateOptions || {}),
+      }
+    }
+
+    if (options.apiUrl) {
+      this.apiUrl = options.apiUrl
+    }
+  }
+
+  /* ----- PRIVATE ----- */
+  private getLib(chainId: NetworkID, apiUrl: string): ParaSwap | null {
```

Comment:
shouldn't we use [private class methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) here?

Reply:
ah, yes 👍